### PR TITLE
TST: Refactor slow tests

### PR DESCRIPTION
**Pull request by mroeschke**

### CODE REVIEW

File: pandas/tests/io/parser/test_c_parser_only.py [hunk 1]
```diff
@@ -44,32 +44,6 @@ def test_buffer_overflow(c_parser_only, malformed):
         parser.read_csv(StringIO(malformed))
 
 
-def test_buffer_rd_bytes(c_parser_only):
```
Comment (mroeschke) [inline pandas/tests/io/parser/test_c_parser_only.py:47]: Based on the original issue, this sounded like a PY2 bug specifically so I don't think this needs testing anymore

File: pandas/tests/io/parser/common/test_chunksize.py [hunk 2]
```diff
@@ -162,14 +163,18 @@ def test_chunk_begins_with_newline_whitespace(all_parsers):
 
 
 @pytest.mark.slow
-def test_chunks_have_consistent_numerical_type(all_parsers):
+def test_chunks_have_consistent_numerical_type(all_parsers, monkeypatch):
+    # mainly an issue with the C parser
+    heuristic = 2**3
     parser = all_parsers
-    integers = [str(i) for i in range(499999)]
+    integers = [str(i) for i in range(heuristic - 1)]
```
Comment (WillAyd) [inline pandas/tests/io/parser/common/test_chunksize.py:170]: Does this need to expand all the way to 2**20?

File: pandas/_libs/parsers.pyx [hunk 3]
```diff
@@ -118,6 +118,8 @@ cdef:
     float64_t NEGINF = -INF
     int64_t DEFAULT_CHUNKSIZE = 256 * 1024
 
+DEFAULT_BUFFER_HEURISTIC = 2 ** 20
```
Comment (WillAyd) [inline pandas/_libs/parsers.pyx:121]: Can this be set as a property of the TextReader? It is a bit ambiguous in a pyx file but in the global namespace all caps I would expect this to be a compile time constant; attaching it as a property makes things clearer

Comment (WillAyd) [reply pandas/_libs/parsers.pyx:121]: Or if this is just for testing maybe you can patch `buffer_lines` directly? The naming here is a bit unclear when scoped outside of the initializer

File: pandas/tests/io/parser/common/test_chunksize.py (hunk 2 above)
Comment (mroeschke) [reply pandas/tests/io/parser/common/test_chunksize.py:170]: I think it needs to hit `DEFAULT_BUFFER_HEURISTIC - 1` to make read_csv with the c parser create a buffer with int and float strings

File: pandas/_libs/parsers.pyx (hunk 3 above)
Comment (mroeschke) [reply pandas/_libs/parsers.pyx:121]: Do you have the same concern about `DEFAULT_CHUNKSIZE` above too?

Ideally I think these magic numbers should at least be made obvious so they can be configured or removed https://github.com/pandas-dev/pandas/issues/53781 and it being in the TextReader would make that less obvious?

Comment (WillAyd) [reply pandas/_libs/parsers.pyx:121]: Unless I am mistaken with how Cython generates this comparing this to `DEFAULT_CHUNKSIZE` is exactly the problem; that is a compile time constant whereas this value can be modified at runtime, but they both look the same

File: pandas/tests/io/parser/common/test_chunksize.py (hunk 2 above)
Comment (WillAyd) [reply pandas/tests/io/parser/common/test_chunksize.py:170]: My mistake - I missed the monkeypatch here so thought this was doing the whole range in 2**20. This looks good

File: pandas/_libs/parsers.pyx (hunk 3 above)
Comment (WillAyd) [reply pandas/_libs/parsers.pyx:121]: Hmm does declaring it `cpdef` help at all? Not worth going down a rabbit hole over if its a hold up

Comment (mroeschke) [reply pandas/_libs/parsers.pyx:121]: Here are my unsuccessful attempts so far just to double check

```
    cdef public:
        int64_t leading_cols, table_width, DEFAULT_BUFFER_HEURISTIC=2**20
                                          ^
------------------------------------------------------------

pandas/_libs/parsers.pyx:365:43: Cannot assign default value to fields in cdef classes, structs or unions
```

```
    cpdef DEFAULT_BUFFER_HEURISTIC=2**20
                                 ^
------------------------------------------------------------

pandas/_libs/parsers.pyx:377:34: Cannot assign default value to fields in cdef classes, structs or unions
```

Comment (WillAyd) [reply pandas/_libs/parsers.pyx:121]: Cool thanks for checking. Let's not get hung up on it for now then - I think just a wart between the C/Python and how they get expressed in the Cython global namespace. Can always come back and refactor if we establish a better pattern

**Review by WillAyd — APPROVED** [review]
lgtm

File: pandas/_libs/parsers.pyx (hunk 3 above)
Comment (mroeschke) [reply pandas/_libs/parsers.pyx:121]: Thanks for confirming!

**Pull request merged**